burn-core/src/nn/sin_embedding.rs

```diff
@@ -0,0 +1,198 @@
+/// the sine and cosine functions, resulting in two arrays of the same shape. These two arrays
+/// are then concatenated along the second dimension, resulting in an array of shape
+/// (sequence_length, channels).
+///
+/// The reference implementation is taken from [Whisper's sinusoids function](
+/// https://github.com/openai/whisper/blob/f572f2161ba831bae131364c3bffdead7af6d210/whisper/model.py#L53).
+///
+///
+/// # Arguments
+///
+/// * `length` - The length of the sequence
+/// * `d_model` - The number of channels
+/// * `max_timescale` - The maximum timescale
+pub fn generate_sinusoids<B: Backend>(
+    length: usize,
+    d_model: usize,
+    max_timescale: usize,
+) -> Tensor<B, 2> {
+    // input checks
+    assert!(d_model % 2 == 0, "d_model must be even");
+    assert!(
+        max_timescale >= length,
+        "max_timescale must be greater than length"
+    );
+
+    // calculate log_timescale_increment by taking the natural logarithm of max_timescale
+    // and dividing it by d_model / 2 - 1.
+    let log_timescale_increment = logf(max_timescale as f32) / ((d_model / 2) - 1) as f32;
+
+    // create an array of inverse timescales, inv_timescales, by generating a sequence of numbers
+    // from 0 to d_model / 2 - 1, multiplying each by -log_timescale_increment, and applying
+    // the exponential function to each.
+    let inv_timescales = (0..d_model / 2)
+        .map(|x| expf(-log_timescale_increment * x as f32))
+        .collect::<Vec<_>>();
+
+    // The scaled_time is calculated by multiplying inv_timescales with an array of numbers
+    // from 0 to length - 1. This operation performs element-wise multiplication between
+    // each number in the sequence from 0 to length - 1 and the inv_timescales array.
+    // This results in a 2D array where each column is a sequence of numbers from 0 to length - 1,
+    // scaled by a different factor from inv_timescales.
+    let scaled_time = (0..length)
+        .map(|i| {
+            inv_timescales
+                .iter()
+                .map(|x| x * i as f32)
+                .collect::<Vec<_>>()
+        })
+        .collect::<Vec<_>>();
+
+    // Calculate the sine and cosine of each element in the scaled_time array,
+    // resulting in two arrays of the same shape.
+    let scaled_time_sin = scaled_time
+        .iter()
+        .map(|x| x.iter().map(|y| sinf(*y)).collect::<Vec<_>>())
+        .collect::<Vec<_>>();
+
+    let scaled_time_cos = scaled_time
+        .iter()
+        .map(|x| x.iter().map(|y| cosf(*y)).collect::<Vec<_>>())
+        .collect::<Vec<_>>();
+
+    // concatenate these two arrays along the second dimension
+    // (i.e., it adds the second array as new columns to the first array)
+    let scaled_time_sin_cos = scaled_time_sin
+        .into_iter()
+        .zip(scaled_time_cos.into_iter())
+        .map(|(mut x, mut y)| {
+            x.append(&mut y);
+            x
+        })
+        .collect::<Vec<_>>();
```

Comment:
Would it be better to use tensor operations here instead? I think the arange function might be missing for the float tensor, but other than that it should be fine.

Reply:
It's easy to understand if the logic is in Rust and there is no advantage to use tensor ops for one time initialization. I'd suggest we'd leave it as is.